convex/util/ollama.ts

```diff
@@ -0,0 +1,154 @@
+import { Ollama } from 'langchain/llms/ollama';
+import {
+  ChatCompletionContent,
+  CreateChatCompletionRequest,
+  LLMMessage,
+  retryWithBackoff,
+} from './openai';
+import { IterableReadableStream } from 'langchain/dist/util/stream';
+
+const ollamaModel = process.env.OLLAMA_MODEL || 'llama2';
+export const UseOllama = process.env.OLLAMA_HOST !== undefined;
+
+// Overload for non-streaming
+export async function ollamaChatCompletion(
+  body: Omit<CreateChatCompletionRequest, 'model'> & {
+    model?: string;
+  } & {
+    stream?: false | null | undefined;
+  },
+): Promise<{ content: string; retries: number; ms: number }>;
+// Overload for streaming
+export async function ollamaChatCompletion(
+  body: Omit<CreateChatCompletionRequest, 'model'> & {
+    model?: string;
+  } & {
+    stream?: true;
+  },
+): Promise<{ content: OllamaCompletionContent; retries: number; ms: number }>;
+export async function ollamaChatCompletion(
+  body: Omit<CreateChatCompletionRequest, 'model'> & {
+    model?: string;
+  },
+) {
+  body.model = body.model ?? 'llama2';
+  const {
+    result: content,
+    retries,
+    ms,
+  } = await retryWithBackoff(async () => {
+    console.log('#### Ollama api ####, using ', ollamaModel);
+
+    const stop = typeof body.stop === 'string' ? [body.stop] : body.stop;
+    const ollama = new Ollama({
+      model: ollamaModel,
+      baseUrl: process.env.OLLAMA_HOST,
+      stop,
+    });
+    const prompt = body.messages.map((m) => m.content).join('\n');
+    console.log('body.prompt', prompt);
+    const stream = await ollama.stream(prompt, { stop });
+    if (body.stream) {
+      return new OllamaCompletionContent(stream, stop ?? []);
+    }
+    let ollamaResult = '';
+    for await (const chunk of stream) {
+      ollamaResult += chunk;
+    }
+    console.log('#### ollama result = ');
+    console.log(ollamaResult);
+    return ollamaResult;
+  });
+
+  return {
+    content,
+    retries,
+    ms,
+  };
+}
+
+// Checks whether a suffix of s1 is a prefix of s2. For example,
+// ('Hello', 'Kira:') -> false
+// ('Hello Kira', 'Kira:') -> true
+const suffixOverlapsPrefix = (s1: string, s2: string) => {
+  for (let i = 1; i <= Math.min(s1.length, s2.length); i++) {
+    const suffix = s1.substring(s1.length - i);
+    const prefix = s2.substring(0, i);
+    if (suffix === prefix) {
+      return true;
+    }
+  }
+  return false;
+};
+
+export class OllamaCompletionContent {
+  private readonly body: IterableReadableStream<string>;
+  private readonly stopWords: string[];
+
+  constructor(body: IterableReadableStream<string>, stopWords: string[]) {
+    this.body = body;
+    this.stopWords = stopWords;
+  }
+
+  // stop words don't always work.
```

Comment:
I think this bug should be fixed now unless there is a bug in the langchain implementation also, you can specify the `stop` words in the langchain ollama call options.